[WIP] Add support for sbroker

fishcakez opened the pull request:
Work in progress :).

Can we drop `stop/1`? It is awkward because it does not work in a supervision tree and as #706 shows we didn't test (or use?) it. I have removed it in a separate commit for now.

Todo:

- [x] Replace use of `Connection` with `GenServer` so not to add an extra dep, was useful to get started though.
- [x] Publish the version of `:sbroker` used and update dep/mix.lock

josevalim commented:
Yes, let's remove `stop/1`.

josevalim commented on integration_test/pool/pool_transaction.exs:
```diff
@@ -6,8 +6,15 @@ defmodule Ecto.Integration.PoolTransactionTest do
+  setup context do
+    case = context[:case]
+    test = context[:test]
+    {:ok, [pool: Module.concat(case, test)]}
```
Do we need this for sbroker or something?

fishcakez replied:
The broker needs a name so the workers can message it.

josevalim commented on lib/ecto/adapters/sojourn_broker/worker.ex:
```diff
@@ -0,0 +1,230 @@
+  end
+
+  def connect(_, %{conn: nil, module: module, params: params,
+  min_backoff: min_backoff} = s) do
```
You can have everything on the same line up to 80 chars. If it is too big, you can do the pattern matching inside the fun itself. :D

josevalim replied:
Same here regarding pattern matching. :)

josevalim commented on mix.exs:
```diff
@@ -3,7 +3,7 @@ defmodule Ecto.Mixfile do
   @version "0.13.0-dev"
   @adapters [:pg, :mysql]
-  @pools [:poolboy]
+  @pools [:poolboy, :sojourn_broker]
```
This means the integration tests will test each pool twice. Well, the unit tests too. I am not sure how to fix it, so you can let it be, I will think about some solutions.

josevalim commented on mix.exs:
```diff
@@ -34,6 +34,8 @@ defmodule Ecto.Mixfile do
 
   defp deps do
     [{:poolboy, "~> 1.4"},
+     {:sbroker, "~> 0.6", github: "fishcakez/sbroker", branch: "gen"},
```
`sbroker` and `connection` should be optional. We will figure out how to make poolboy optional later (please add a TODO on top of the poolboy one).

josevalim replied:
Btw is the `connection` module meant to be used exclusively by sbroker?

fishcakez replied:
It is for any connection process and nothing to do with `sbroker`. The mongo driver uses it for example https://github.com/ericmj/mongodb/blob/08b63a1aaca3bbf8f8f48daa90bc83630d46455f/lib/mongo/connection.ex#L10. @whatyouhide did a nice blog post about using it: http://andrealeopardi.com/posts/handling-tcp-connections-in-elixir/.

I am keen to remove from here because it would be an optional dep and the would be confusing. Also there could be a version mismatch with a driver.

josevalim replied:
Ok, your call. If you think it would be better to use it with *everything* that is fine for me too.

josevalim replied:
OR even push postgrex and mariaex and friends to use it too.

whatyouhide replied:
Not part of this discussion, but FWIW I'm using it on my OrientDB adapter as well :)

fishcakez replied:
> OR even push postgrex and mariaex and friends to use it too.

Before we do that I think we should see if the "smart" NoSQL connections perform better than the "dumb" SQL connections under network partition. What's good for `poolboy` might be bad for `sbroker` and vice versa.

josevalim replied:
Oops, I meant the Connection module, not sbroker itself. :)

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

fishcakez replied:
Have remove `Connection`. Still need to publish a new version of `sbroker`.

fishcakez commented:
I've rebased this and:
* Made `break/3` a cast for `sbroker` (blocking not required as worker checks in)
* Refactored the worker to a `GenServer` to remove the `Connection` dependency

fishcakez commented on integration_test/cases/pool.exs:
```diff
@@ -16,9 +16,7 @@ defmodule Ecto.Integration.PoolTest do
-    {_, pool_name, _} = PoolRepo.__pool__
-    pool_pid = Process.whereis(pool_name)
-    assert {:error, {:already_started, ^pool_pid}} = PoolRepo.start_link()
+    assert {:error, {:already_started, _}} = PoolRepo.start_link()
```
Must this point to the pool process? In this branch a supervisor (which is registered as `Module.concat(pool_name, Supervisor)`)

josevalim replied:
No, it does not. :)

josevalim replied:
As long as the pool name is the one coming from the `__pool__` tuple, we are good to go.

fishcakez commented on integration_test/pool/pool_run.exs:
```diff
@@ -51,11 +61,12 @@ defmodule Ecto.Integration.PoolRunTest do
     end
 
-    refute Process.alive?(conn)
+    assert_receive {:DOWN, ^monitor, _, _, _}
```
break is asynchronous

fishcakez commented on lib/ecto/adapters/sojourn_broker.ex:
```diff
@@ -0,0 +1,124 @@
+  @doc """
+  Stop the pool.
+  """
+  def stop(pool) do
```
this should go as `stop/1` is gone.

josevalim replied:
So just remove it? :)

fishcakez replied:
It was a reminder for myself :P.

josevalim commented:
@fishcakez this looks great to me. Let's just release a new sbroker and get going? I want to clean up the pool structure and tests, but I will do that in another PR. :)

fishcakez commented:
ugh, we need to make this lazy, otherwise we hit the same issue as the sandbox. I'll do this after lunch.

fishcakez commented:
I added lazy connect and tests. Implementation is clean if only the first connect is lazy - which is all we need.
